[TorchScript] C++ interface of to_<backend>

iseeyuan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#51797 [TorchScript] C++ interface of to_<backend>**

The C++ API, ```codegen_backend_module``` is added to ```to_<backend>```. Python related stuffs are decoupled in this function. It can be used from both C++ and python. 

* Tests
Python: The existing ```test_backends.py```, which calls the C++ API under the hood.
C++: The end-to-end test of ```jit.BackendTest.ToBackend``` is added in ```test_backend.cpp```. The original class definitions in this file is moved to ```test_backend_lib.cpp```

Differential Revision: [D26280518](https://our.internmc.facebook.com/intern/diff/D26280518/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D26280518/)!

iseeyuan committed:
[TorchScript] C++ interface of to_<backend>



Differential Revision: [D26280518](https://our.internmc.facebook.com/intern/diff/D26280518/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D26280518/)!

[ghstack-poisoned]

iseeyuan committed:
Update on "[TorchScript] C++ interface of to_<backend>"


Differential Revision: [D26280518](https://our.internmc.facebook.com/intern/diff/D26280518/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D26280518/)!

[ghstack-poisoned]

iseeyuan committed:
Update on "[TorchScript] C++ interface of to_<backend>"


Differential Revision: [D26280518](https://our.internmc.facebook.com/intern/diff/D26280518/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D26280518/)!

[ghstack-poisoned]

iseeyuan committed:
Update on "[TorchScript] C++ interface of to_<backend>"


Differential Revision: [D26280518](https://our.internmc.facebook.com/intern/diff/D26280518/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D26280518/)!

[ghstack-poisoned]

iseeyuan committed:
Update on "[TorchScript] C++ interface of to_<backend>"


Differential Revision: [D26280518](https://our.internmc.facebook.com/intern/diff/D26280518/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D26280518/)!

[ghstack-poisoned]

iseeyuan commented on torch/csrc/jit/backends/backend_detail.h:
```diff
@@ -63,6 +63,11 @@ TORCH_API bool hasBackendPreprocessFunction(const std::string& name);
 TORCH_API BackendPreprocessFunction
```
Right, they are referenced by codegen_backend_module

iseeyuan committed:
Update on "[TorchScript] C++ interface of to_<backend>"


The C++ API, ```codegen_backend_module``` is added to ```to_<backend>```. Python related stuffs are decoupled in this function. It can be used from both C++ and python. 

* Tests
Python: The existing ```test_backends.py```, which calls the C++ API under the hood.
C++: The end-to-end test of ```jit.BackendTest.ToBackend``` is added in ```test_backend.cpp```. The original class definitions in this file is moved to ```test_backend_lib.cpp```

Differential Revision: [D26280518](https://our.internmc.facebook.com/intern/diff/D26280518/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D26280518/)!

[ghstack-poisoned]

raziel commented on torch/csrc/jit/backends/backend_detail.h:
```diff
@@ -58,11 +58,16 @@ TORCH_API void registerBackendPreprocessFunction(
+bool hasBackendPreprocessFunction(const std::string& name);
 
-TORCH_API BackendPreprocessFunction
+BackendPreprocessFunction
```
I think you can move this to the same line (otherwise CLANFORMAT will likely complain).

iseeyuan committed:
Update on "[TorchScript] C++ interface of to_<backend>"


The C++ API, ```codegen_backend_module``` is added to ```to_<backend>```. Python related stuffs are decoupled in this function. It can be used from both C++ and python. 

* Tests
Python: The existing ```test_backends.py```, which calls the C++ API under the hood.
C++: The end-to-end test of ```jit.BackendTest.ToBackend``` is added in ```test_backend.cpp```. The original class definitions in this file is moved to ```test_backend_lib.cpp```

Differential Revision: [D26280518](https://our.internmc.facebook.com/intern/diff/D26280518/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D26280518/)!

[ghstack-poisoned]

iseeyuan committed:
Update on "[TorchScript] C++ interface of to_<backend>"


The C++ API, ```codegen_backend_module``` is added to ```to_<backend>```. Python related stuffs are decoupled in this function. It can be used from both C++ and python. 

* Tests
Python: The existing ```test_backends.py```, which calls the C++ API under the hood.
C++: The end-to-end test of ```jit.BackendTest.ToBackend``` is added in ```test_backend.cpp```. The original class definitions in this file is moved to ```test_backend_lib.cpp```

Differential Revision: [D26280518](https://our.internmc.facebook.com/intern/diff/D26280518/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D26280518/)!

[ghstack-poisoned]

raziel reviewed (commented):
LGTM

I'll leave it to @SplitInfinity to approve.

SplitInfinity commented on test/cpp/jit/CMakeLists.txt:
```diff
@@ -74,7 +75,7 @@ if(USE_SYSTEM_ONNX)
   target_link_libraries(test_jit PRIVATE onnx_proto onnx)
 endif()
 
-set(JIT_TEST_DEPENDENCIES torch gtest)
```
What's the reason for this change? Is it because `jitbackend_test` includes `torch`?

iseeyuan replied:
Yes. I'm trying to resolve CI failures and testing if this change could help. It turns out not. I may change it back after verification.

SplitInfinity commented on test/cpp/jit/test_backend.cpp:
```diff
@@ -1,81 +1,42 @@
+#include <torch/csrc/jit/api/module.h>
 #include <torch/csrc/jit/backends/backend.h>
+#include <torch/csrc/jit/backends/backend_detail.h>
+#include <torch/csrc/jit/frontend/resolver.h>
```
Why is this header needed? I don't see any of the types from the file used here.

iseeyuan replied:
Removed.

SplitInfinity commented on test/cpp/jit/test_backend.cpp:
```diff
@@ -1,81 +1,42 @@
 #include <torch/csrc/jit/backends/backend.h>
+#include <torch/csrc/jit/backends/backend_detail.h>
+#include <torch/csrc/jit/frontend/resolver.h>
+#include <torch/torch.h>
```
Why is this header needed? I'm not familiar with it so it might be used, I'm just curious.

iseeyuan replied:
It's a header of all file, probably for ease of use, especially for generated headers? The tensor factory methods (torch::ones) are from what this header includes.

SplitInfinity commented on test/cpp/jit/test_backend_lib.cpp:
```diff
@@ -0,0 +1,81 @@
+// necessary to test that the JIT backend registration endpoints and
+// code generation are working correctly. It is not intended to
+// produce numerically correct results.
+class TORCH_API TestBackend : public PyTorchBackendInterface {
```
As I mentioned before, removing `TORCH_API` might fix your test failures.

SplitInfinity commented on torch/csrc/jit/backends/backend_detail.h:
```diff
@@ -58,11 +58,15 @@ TORCH_API void registerBackendPreprocessFunction(
-TORCH_API bool hasBackendPreprocessFunction(const std::string& name);
+bool hasBackendPreprocessFunction(const std::string& name);
 
-TORCH_API BackendPreprocessFunction
-getBackendPreprocessFunction(const std::string& name);
+BackendPreprocessFunction getBackendPreprocessFunction(const std::string& name);
```
Why is `TORCH_API` no longer needed? Is it because these functions are used only in `libtorch` after moving the codegen to `backend_detail.cpp`, and no longer used in `libtorch_python`?

raziel replied:
Yep, I originally added the TORCH_API because of the split but now shouldn't be needed.

Martin can confirm that's the case now.

iseeyuan replied:
It's referenced by codegen_backend_module, which is already marked as ```TORCH_API```

SplitInfinity reviewed (approved):
Changes look good, but please fix CI before merging.

SplitInfinity commented on torch/csrc/jit/backends/backend_init.cpp:
```diff
@@ -118,6 +118,19 @@ void toBackendSelectiveImpl(
+      orig_module,
+      toIValue(method_compile_spec, any_dict_ty).toGenericDict(),
+      any_dict_ty);
+};
```
```suggestion
}
```

SplitInfinity commented on torch/csrc/jit/backends/backend_init.cpp:
```diff
@@ -118,6 +118,19 @@ void toBackendSelectiveImpl(
   }
 }
 
+Module codegen_lambda(
```
I would rename this now that it is no longer a lambda.

iseeyuan committed:
Update on "[TorchScript] C++ interface of to_<backend>"


The C++ API, ```codegen_backend_module``` is added to ```to_<backend>```. Python related stuffs are decoupled in this function. It can be used from both C++ and python. 

* Tests
Python: The existing ```test_backends.py```, which calls the C++ API under the hood.
C++: The end-to-end test of ```jit.BackendTest.ToBackend``` is added in ```test_backend.cpp```. The original class definitions in this file is moved to ```test_backend_lib.cpp```

Differential Revision: [D26280518](https://our.internmc.facebook.com/intern/diff/D26280518/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D26280518/)!

[ghstack-poisoned]

iseeyuan committed:
Update on "[TorchScript] C++ interface of to_<backend>"


The C++ API, ```codegen_backend_module``` is added to ```to_<backend>```. Python related stuffs are decoupled in this function. It can be used from both C++ and python. 

* Tests
Python: The existing ```test_backends.py```, which calls the C++ API under the hood.
C++: The end-to-end test of ```jit.BackendTest.ToBackend``` is added in ```test_backend.cpp```. The original class definitions in this file is moved to ```test_backend_lib.cpp```

Differential Revision: [D26280518](https://our.internmc.facebook.com/intern/diff/D26280518/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D26280518/)!

[ghstack-poisoned]

malfet commented on test/cpp/jit/CMakeLists.txt:
```diff
@@ -74,7 +75,7 @@ if(USE_SYSTEM_ONNX)
 endif()
 
-set(JIT_TEST_DEPENDENCIES torch gtest)
+set(JIT_TEST_DEPENDENCIES jitbackend_test gtest)
```
On Linux, linker would not add a dependency for shared library, unless target actually depends on symbol from that library.
One way to workaround it is to add something like:
```
+if(LINUX)
  #Update to target_link_options when CMake version can be upgraded
  target_link_libraries(test_jit PRIVATE "-Wl,--no-as-needed,$<TARGET_FILE:jitbackend_test>")
endif()
```

iseeyuan replied:
Thanks @malfet !

iseeyuan committed:
Update on "[TorchScript] C++ interface of to_<backend>"


The C++ API, ```codegen_backend_module``` is added to ```to_<backend>```. Python related stuffs are decoupled in this function. It can be used from both C++ and python. 

* Tests
Python: The existing ```test_backends.py```, which calls the C++ API under the hood.
C++: The end-to-end test of ```jit.BackendTest.ToBackend``` is added in ```test_backend.cpp```. The original class definitions in this file is moved to ```test_backend_lib.cpp```

Differential Revision: [D26280518](https://our.internmc.facebook.com/intern/diff/D26280518/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D26280518/)!

[ghstack-poisoned]

raziel commented on torch/custom_class_detail.h:
```diff
@@ -135,7 +135,6 @@ inline void checkValidIdent(const std::string& str, const char *type) {
 }
 
 } // namespace detail
-
```
maybe revert this change?

iseeyuan replied:
Good catch. Just updated it, and also rebase to internal fbobjc/stable for easier landing.

iseeyuan committed:
Update on "[TorchScript] C++ interface of to_<backend>"


The C++ API, ```codegen_backend_module``` is added to ```to_<backend>```. Python related stuffs are decoupled in this function. It can be used from both C++ and python. 

* Tests
Python: The existing ```test_backends.py```, which calls the C++ API under the hood.
C++: The end-to-end test of ```jit.BackendTest.ToBackend``` is added in ```test_backend.cpp```. The original class definitions in this file is moved to ```test_backend_lib.cpp```

Differential Revision: [D26280518](https://our.internmc.facebook.com/intern/diff/D26280518/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D26280518/)!

[ghstack-poisoned]
